File: 프로그래머스/lv2/68936. 쿼드압축 후 개수 세기/쿼드압축 후 개수 세기.py
result = []

def solution(arr):
    get_append(arr, 0, len(arr[0]), 0, len(arr))

    ans = [0,0]
    for i in range(0, len(result)):
        ans[result[i]]+=1
        
    return ans

def zipping(matrix, start_x, end_x, start_y, end_y):
    half_x = int((end_x + start_x) / 2)
    half_y = int((end_y + start_y) / 2)

    get_append(matrix, start_x, half_x, start_y, half_y)
    get_append(matrix, half_x, end_x, start_y, half_y)
    get_append(matrix, start_x, half_x, half_y, end_y)
    get_append(matrix, half_x, end_x, half_y, end_y)

def get_append(matrix, start_x, end_x, start_y, end_y):
    if is_all_same(matrix, start_x, end_x, start_y, end_y):
        result.append(matrix[start_y][start_x])
    else:
        zipping(matrix, start_x, end_x, start_y, end_y)

def is_all_same(matrix, start_x, end_x, start_y, end_y):
    value = matrix[start_y][start_x]
    for y in range(start_y, end_y):
        for x in range(start_x, end_x):
            if value != matrix[y][x]:
                return False
    return True

File: 프로그래머스/lv2/68936. 쿼드압축 후 개수 세기/쿼드압축 후 개수 세기.java
import java.util.*;

class Solution {
    static List<Integer> result;
    public int[] solution(int[][] arr){
        result = new ArrayList<>();
        getAppend(arr, 0, arr[0].length, 0, arr.length);
        int[] answer = {0,0};

        for (Integer resultInteger : result) {
            answer[resultInteger]++;
        }
        return answer;
    }
    public void getAppend(int[][] matrix, int startX, int endX, int startY, int endY){
        if(isAllSame(matrix, startX, endX, startY, endY)){
            result.add(matrix[startY][startX]);
        }else{
            zipping(matrix, startX, endX, startY, endY);
        }
    }
    public void zipping(int[][] matrix, int startX, int endX, int startY, int endY){
        int halfX = (endX + startX) / 2;
        int halfY = (endY + startY) / 2;

        getAppend(matrix, startX, halfX, startY, halfY);
        getAppend(matrix, halfX, endX, startY, halfY);
        getAppend(matrix, startX, halfX, halfY, endY);
        getAppend(matrix, halfX, endX, halfY, endY);
    }
    
    public boolean isAllSame(int[][] matrix, int startX, int endX, int startY, int endY){
        int first = matrix[startY][startX];
        for(int i = startY; i < endY; i++){
            for(int j = startX; j < endX; j++){
                if(first != matrix[i][j]){
                    return false;
                }
            }
        }
        return true;
    }
}    